src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -51,6 +57,8 @@ class RsInferenceResult(
 class RsInferenceContext {
     private val bindings: MutableMap<RsPatBinding, Ty> = HashMap()
     private val exprTypes: MutableMap<RsExpr, Ty> = HashMap()
+    private val resolvedPaths: MutableMap<RsPathExpr, List<RsCompositeElement>> = HashMap()
+    private val pathRefinements: MutableList<Pair<RsPathExpr, TraitRef>> = mutableListOf()
```

Comment:
I don't sure in `refinement` name. It's something like additional info for resolved path that should be used after type inference to provide more precise path (not to trait but to impl)